libs/sdk-core/src/lsps0/transport.rs

```diff
@@ -0,0 +1,677 @@
+                        }
+                    } else {
+                        debug!("LSPS0: got error without listener: {:?}", e)
+                    }
+                }
+            }
+        }
+    }
+
+    pub async fn request_response<TRequest, TResponse>(
+        &self,
+        method: String,
+        peer_id: Vec<u8>,
+        req: &TRequest,
+        timeout: Duration,
+    ) -> Result<TResponse, Error>
+    where
+        TRequest: serde::Serialize,
+        TResponse: serde::de::DeserializeOwned,
+    {
+        let request_id = generate_request_id();
+        let wrapped_req = JsonRpcRequest {
+            id: request_id.clone(),
+            jsonrpc: String::from(JSONRPC_VERSION),
+            method,
+            params: req,
+        };
+        let payload = serde_json::to_vec(&wrapped_req)?;
+        let (tx, rx) = oneshot::channel();
+        let handler_id = get_request_handler_id(&request_id, peer_id.clone());
+        (*self.response_handlers.lock().await).insert(handler_id.clone(), tx);
+
+        if let Err(e) = self
+            .node
+            .send_custom_message(CustomMessage {
+                peer_id,
+                message_type: LSPS0_MESSAGE_TYPE,
+                payload,
+            })
+            .await
+        {
+            (*self.response_handlers.lock().await).remove(&handler_id);
+            return Err(e.into());
+        }
+
+        let result_or = tokio::time::timeout(timeout, rx).await?;
+        let response_or_error = result_or?;
+        if let Some(response) = response_or_error.response {
+            let resp = serde_json::from_value::<TResponse>(response.result)?;
+            Ok(resp)
+        } else if let Some(error) = response_or_error.error {
+            Err(Error::Remote {
+                code: error.error.code,
+                message: error.error.message,
+                data: error.error.data,
+            })
+        } else {
+            Err(Error::Local(anyhow!("did not get response or error")))
+        }
+    }
+
+    pub async fn stream_notifications<TNotification>(
+        &self,
+        method: String,
+        node_id: Vec<u8>,
+    ) -> Result<mpsc::Receiver<TNotification>, Error>
+    where
+        TNotification: serde::de::DeserializeOwned + std::marker::Send + 'static,
+    {
+        let (tx, rx) = mpsc::channel(100);
+        let id = get_notification_handler_id(&method, node_id);
+        (*self.notification_handlers.write().await)
+            .insert(id, Box::new(NotificationHandler::<TNotification> { tx }));
+
+        Ok(rx)
+    }
+}
+
+fn generate_request_id() -> String {
+    Alphanumeric.sample_string(&mut rand::thread_rng(), 21)
+}
+
+fn get_request_handler_id(request_id: &str, node_id: Vec<u8>) -> String {
+    let mut id = hex::encode(node_id);
+    id.push('|');
+    id.push_str(request_id);
+    id
+}
+
+fn get_notification_handler_id(method: &str, node_id: Vec<u8>) -> String {
```

Comment:
Shouldn't we further scope the notification handler by a random request id? I might be missing something but it looks like that if the client invokes the same method in parallel  the later invocation will replace the the entry in the response_handlers of the former that will hang as a result?

Reply:
It works like that for request/response. Notifications don't have an id, just a method. Notifications are 'streaming' from the server side.

Reply:
Great! I guess I confused it.